README.md
# couchsc
Relax. Domain model based persistence made easy.

CouchSC puts a domain model based REST-wrapper around CouchDB for simple persistence of domain model instances.

# Get it somehow running.

## Preconditions---CouchDB
0. Download a recent copy of CouchDB (https://couchdb.apache.org), install and start it.
0. You need Java 8.

## Precondition---ProcessEditor with WebModeler
0. Download the recent copy of https://github.com/frapu78/processeditor.
0. Build the jar-server executable following the documentation.
0. Start the jar-server executable and browse to http://localhost:1205
0. Login to the WebModeler (default credentials: root/inubit).
0. Import resources/fuhrpark.model.

## Running it!
0. Create a project in an IDE of your choice, add the content of this repository as well as the processeditor.jar file.
0. Open the imported model "Fuhrpark" in the WebModeler and copy the URL, e.g. http://localhost:1205/models/1520138062 - Last id will be different!
0. Open the CouchSCServer source and configure the modelLocation variable accordingly to the copied URL (as a source for the model to be used).
0. Configure the couchdb credentials.
0. Run the CouchSCServer.
0. Point your browser to http://localhost:2107 and relax!




CHANGELOG.txt
Changelog for CouchSC
=====================

Todo
* Check if association key in couchdb can be created with key = "source,target" and can be searched / sorted by target (endswidth)
* FIX support for unicode, e.g. German characters currently lead to an endless-loop(!)

2019-03-10
* Worked on search API
* Introduced HTTPConstants for handler

2019-01-03
* Renamed "type" in "$type" for couchdb

2018-12-27
* Worked on InstanceConnector to support limits, keyed searches
* Reworked domain model to contain abstract "Person" class, added "Kunde", Added Mitarbeiter aggregates Mitarbeiter ("chef_von")

2018-12-21
* Added JQuery resources and image resources to local folder
* Added support for binary files

2018-12-20
* Worked on Associations; view of different target types now supported

2018-12-18
* Added first version of ResourceHandler

2018-05-31
* Introduced JQueryUI

2018-05-30
 * Updated support for newer couchdb versions with basic authentication
 * Introduced support for enumerations via EnumerationClasses and drop-down boxes
 * Started implementation of DefaultRenderer for plain HTML output

src/net/frapu/couchsc/CouchSCServer.java
/**
 * Server package for the Couch Solution Center (CouchSC).
 */
package net.frapu.couchsc;

import com.inubit.research.client.JSONHttpRequest;
import com.inubit.research.client.UserCredentials;
import net.frapu.couchsc.handler.DefaultHandler;
import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.URI;
import java.util.HashMap;
import java.util.Map;
import net.frapu.code.visualization.ProcessEdge;
import net.frapu.code.visualization.ProcessNode;
import net.frapu.code.visualization.ProcessUtils;
import net.frapu.code.visualization.domainModel.Association;
import net.frapu.code.visualization.domainModel.Attribute;
import net.frapu.code.visualization.domainModel.DomainClass;
import net.frapu.code.visualization.domainModel.DomainModel;
import net.frapu.code.visualization.domainModel.DomainUtils;
import net.frapu.couchsc.utils.CouchDBHelper;
import org.json.JSONException;
import org.json.JSONObject;

/**
 *
 * The main class to start the server.
 *
 * @author frank
 */
public class CouchSCServer {

    private HttpServer server;
    private static CouchSCServer instance;
    private boolean running = false;
    private DomainModel model;
    private InstanceConnector ic = null;
    private String couchDBVersion = "";
    private static final int PORT = 2107;

    private static final int LIMIT = 10;

    private final URI modelLocation = URI.create(("http://127.0.0.1:1205/models/1520138062"));
    // Credentials for CouchDB
    private final String couchDBUrl = "http://localhost:5984/";
    private final String couchDBCredentials = CouchDBHelper.getBasicAuthString("admin", "admin");
    // Credentials for ProcessEditorServer
    private final UserCredentials modelUserCredentials = new UserCredentials("http://127.0.0.1:1205", "root", "inubit");

    public static final String INDICATOR = "$";
    public static final String TYPE_INDICATOR = INDICATOR+"type";

    public CouchSCServer() throws Exception {
    }

    public void start() {
        System.out.println("Server started at "+server.getAddress()+"...");
        if (this.server != null) {
            this.server.start();
        }
        this.running = true;
    }

    /**
     * Stop the server
     */
    public void stop() {
        System.out.println("Server stopped...");
        if (server != null) {
            this.server = null;
        }

        this.running = false;
    }

    public void updateDBViews() {
        try {
            createClassViews();
            createAssocViews();
        } catch (Exception e) {
            System.err.println("ERROR UPDATING DB VIEWS");
            e.printStackTrace();
        }

    }
    
    private void createAssocViews() throws JSONException, IOException {
        Map<String, JSONObject> assocs = new HashMap<String, JSONObject>();
        // Iterate over all edges
        for (ProcessEdge edge : model.getEdges()) {
            if (edge instanceof Association) {
                // Generate Javascript
                JSONObject map = new JSONObject();
                Association assoc = (Association)edge;
                String function = "function(doc) {";
                function += "if( doc."+TYPE_INDICATOR+"==\"Association\") {";
                function += "emit( doc.source, doc); ";
                function += "}}";
                map.put("map", function);
                // Create a view for each class
                JSONObject view = new JSONObject();                
                assocs.put("AssociationsFrom", map);
            }
        }
        JSONObject doc = new JSONObject();
        doc.put("_id", "_design/assoc_" + ic.getDbName());
        doc.put("language", "javascript");
        doc.put("views", assocs);
        try {
            // Try to catch old version's _rev id
            JSONObject ov = ic.getDocument("_design/assoc_" + ic.getDbName());
            doc.put("_rev", ov.get("_rev"));
        } catch (Exception e) {
        }
        JSONObject result = ic.putDocument(doc);
    }    

    private void createClassViews() throws JSONException, IOException {
        Map<String, JSONObject> views = new HashMap<String, JSONObject>();
        // Iterate over all classes
        for (ProcessNode node : model.getNodes()) {
            if (node instanceof DomainClass) {
                // Generate Javascript
                JSONObject map = new JSONObject();
                DomainClass dc = (DomainClass) node;
                String function = "function(doc) {";
                function += "if(";
                // Iterate over all child classes, too
                int size = DomainUtils.getChildren(dc, model).size();
                for (DomainClass dc1 : DomainUtils.getChildren(dc, model)) {
                    function += "doc."+TYPE_INDICATOR+"==\"" + InstanceConnector.normalize(dc1.getName()) + "\"";
                    if ((size--) > 1) {
                        function += "|";
                    }
                }
                function += ") {";
                function += "emit(";
                // Check if the DomainClass has a key
                if (dc.getKey() != null) {
                    function += dc.getKey();
                } else {
                    // Use id as key
                    function += "doc._id";
                }
                function += ",";
                if (((DomainClass) node).getAttributesByIDs().size() > 0) {
                    function += "{";
                    int count = 0;
                    for (Attribute a : dc.getAttributesByIDs().values()) {
                        function += a.getName() + ": doc." + a.getName();
                        if (count < dc.getAttributesByIDs().size() - 1) {
                            function += ", ";
                        }
                        count++;
                    }
                    function += "}";
                } else {
                    function += " null";
                }
                function += ");";
                function += "}}";
                map.put("map", function);
                // Create a view for each class
                JSONObject view = new JSONObject();
                views.put(InstanceConnector.normalize(node.getName()), map);
            }
        }
        JSONObject doc = new JSONObject();
        doc.put("_id", "_design/view_" + ic.getDbName());
        doc.put("language", "javascript");
        doc.put("views", views);
        try {
            // Try to catch old version's _rev id
            JSONObject ov = ic.getDocument("_design/view_" + ic.getDbName());
            doc.put("_rev", ov.get("_rev"));
        } catch (Exception e) {
        }
        JSONObject result = ic.putDocument(doc);
    }

    public boolean isRunning() {
        return this.running;
    }

    public URI getRecentModelLocation() {
        return modelLocation;
    }

    public UserCredentials getRecentUserCredentials() {
        return modelUserCredentials;
    }

    public void init() throws Exception {
        // Load sample model
        model = (DomainModel) ProcessUtils.parseProcessModelSerialization(modelLocation,
                modelUserCredentials);
        // Connect to CouchDB and get version (just a check)
        JSONHttpRequest req = new JSONHttpRequest(URI.create(couchDBUrl));
        req.setRequestProperty("Authorization", couchDBCredentials);
        JSONObject obj = req.executeGetRequest();
        couchDBVersion = obj.optString("version");
        // Establish instance connector
        ic = new InstanceConnector(model, couchDBUrl, couchDBCredentials);
        // Create default entry

        /*
        JSONObject doc = new JSONObject();
        doc.put("_id", "123-456-788");
        doc.put("type", "Person");
        doc.put("Vorname", "Frank");
        doc.put("Nachname", "Puhlmann");
        doc.put("Geburtsdatum", "15/03/1978");

        System.out.println(doc);
        try {
            ic.putDocument(doc);
        } catch (IOException ioe) {
            // Don't care if already exists
        }
        */

        // Update db views
        System.out.print("Updating db views...");
        updateDBViews();
        System.out.println("OK");

        // Bind to all interfaces
        InetSocketAddress address = new InetSocketAddress(PORT);
        this.server = HttpServer.create(address, 255);
        // Add contexts
        server.createContext("/", new DefaultHandler());
    }

    public DomainModel getDomainModel() {
        return model;
    }

    public String getCouchDB() {
        return couchDBUrl;
    }

    public String getCouchDBVersion() {
        return couchDBVersion;
    }

    // Returns the limit for each entry to show
    public int getLimit() { return LIMIT;}

    public static CouchSCServer getInstance() {
        if (instance == null) {
            try {
                instance = new CouchSCServer();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return instance;
    }

    public InstanceConnector getInstanceConnector() {
        return ic;
    }

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) throws Exception {
        CouchSCServer server = getInstance();
        server.init();
        server.start();

    }
}
